Redesign survey multiple choice

AlexSCorey opened the pull request:
##### SUMMARY
Addresses #6464.  

This new design improves UI and reduces the risk to having mismatching choices and default values. 
##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
 - UI

##### AWX VERSION


##### ADDITIONAL INFORMATION
![Screen Shot 2021-04-15 at 3 55 07 PM](https://user-images.githubusercontent.com/39280967/114930609-34a12a80-9e03-11eb-8538-77cb9586a218.png)

AlexSCorey committed:
adding and removing inputs properly

AlexSCorey committed:
some refactoring, and checking and unchecking boxes

AlexSCorey committed:
can write in inputs properly

AlexSCorey committed:
Adds styling, and dynamic rendering of extra fields

nixocio commented:
hey @AlexSCorey, I got this code locally, and I switched from `multiple select` to `single select`, and I got the following errors:

1 - Empty default seems to be reason of this error.

<img width="527" alt="image" src="https://user-images.githubusercontent.com/9053044/115030004-98276880-9e94-11eb-9c80-01bdadc45f19.png">


<img width="842" alt="image" src="https://user-images.githubusercontent.com/9053044/115029688-3cf57600-9e94-11eb-8361-927b0c52d86f.png">


2 - 

<img width="1415" alt="image" src="https://user-images.githubusercontent.com/9053044/115030297-f05e6a80-9e94-11eb-9e49-9e5ad016c966.png">

AlexSCorey commented:
Add tooltip on the checkbox

AlexSCorey commented:
recheck

nixocio commented:
For the most part this looks great @AlexSCorey, I got the code changes again, and I want to verify a few scenarios with you.

1 - Survey is null 

<img width="1420" alt="image" src="https://user-images.githubusercontent.com/9053044/115458132-56c7ed80-a1f3-11eb-8ef5-dab66815529e.png">

Attempt to save a Multiple Choice - Single Selection with an empty default, click to select this empty choice as default one and the `survey is null` will show. This error does not disappear after adding more options - and changing the default.

Error does not disappear after adding more options, and changing default one. 

<img width="1418" alt="image" src="https://user-images.githubusercontent.com/9053044/115458214-7232f880-a1f3-11eb-8d81-4836f8a6e3ab.png">

Same error happening to Multiple Choice - Multiple Selection.

<img width="1450" alt="image" src="https://user-images.githubusercontent.com/9053044/115458524-cfc74500-a1f3-11eb-91cf-c4229cd8776a.png">


2 - Being able to save a survey with an unique empty choice, and required - It generates an error when attempting to launch the template.

<img width="1468" alt="image" src="https://user-images.githubusercontent.com/9053044/115461513-84169a80-a1f7-11eb-95d4-789e69a530a8.png">

<img width="1243" alt="image" src="https://user-images.githubusercontent.com/9053044/115461644-a90b0d80-a1f7-11eb-9989-65dd74da2e11.png">

<img width="1275" alt="image" src="https://user-images.githubusercontent.com/9053044/115461916-fc7d5b80-a1f7-11eb-9582-15717b1ffc5c.png">

AlexSCorey committed:
This puts the formatted default and choice values on the formik object.
When we go to submit the form to the api we format it again in a way
the api will recognize.  Allowing formik to manage updating, the choices
and the default values this way cleans up the code and removes a bunch of unnecessary
splitting and joining of the choices an default choices strings

AlexSCorey committed:
updates strings

AlexSCorey commented:
recheck

nixocio commented:
hey @AlexSCorey, I got your latest changes, the help text is a nice addition - very helpful.

But the following issues remains:

1 - Attempt to save an empty choice as default.

<img width="1457" alt="image" src="https://user-images.githubusercontent.com/9053044/116077377-6caa3800-a663-11eb-8ab2-2bd74e649950.png">


2 - Attempt to save a choice without being the default one.

<img width="1460" alt="image" src="https://user-images.githubusercontent.com/9053044/116077707-c579d080-a663-11eb-9ca4-4f5d77e8a542.png">

marshmalien commented on awx/ui_next/src/components/FormField/FormField.jsx:
```diff
@@ -29,7 +29,6 @@ function FormField(props) {
           helperText={helperText}
           helperTextInvalid={meta.error}
           isRequired={isRequired}
-          validated={isValid ? 'default' : 'error'}
```
Did you mean to remove this line? I thought we needed it to set helper text under text area fields. 

This screenshot is from the metadata step of credential plugin prompts. See missing helper text under **Unsigned Public Key**.
<img width="911" alt="Screen Shot 2021-04-26 at 11 49 56 AM" src="https://user-images.githubusercontent.com/15881645/116115101-0be12680-a688-11eb-90e7-275b50065a3f.png">

AlexSCorey replied:
No, didn't mean to do that.

marshmalien commented on awx/ui_next/src/components/FormField/TextAndCheckboxField.jsx:
```diff
@@ -0,0 +1,136 @@
+      {formattedChoicesField.value.map(({ choice, isDefault }, i) => (
+        <InputGroup>
+          <TextInput
+            aria-label={choice}
```
Seeing this console error whenever I add an empty multiple choice option. We should probably set a default label whenever the user adds a fresh multiple choice option. 

```
{
    "name": "formattedChoices",
    "value": [
        {
            "choice": "^check🕴🕴🕴",
            "isDefault": true
        },
        {
            "choice": "",
            "isDefault": false
        }
    ]
}
```

<img width="669" alt="Screen Shot 2021-04-26 at 12 04 46 PM" src="https://user-images.githubusercontent.com/15881645/116116126-cbce7380-a688-11eb-9e18-b5acb044089c.png">

AlexSCorey replied:
I'm reluctant to set a default value for a new choice because I don't want the user to accidentally submit the default choice, but I can set an aria-label

marshmalien commented on awx/ui_next/src/screens/Template/Survey/SurveyQuestionAdd.jsx:
```diff
@@ -9,23 +9,48 @@ export default function SurveyQuestionAdd({ survey, updateSurvey }) {
       }
-      const newSpec = survey.spec ? survey.spec.concat(question) : [question];
+      const newSpec = survey.spec ? survey.spec.concat(formData) : [formData];
+
```
The form allows empty `" "` multiselect options. Is that expected? They show up as empty tags in the survey question list view. I _think_ this is happening in both Add and Edit. 

<img width="224" alt="Screen Shot 2021-04-26 at 12 21 01 PM" src="https://user-images.githubusercontent.com/15881645/116140636-9c7a2f80-a6a5-11eb-928c-5b4199ea1415.png">

AlexSCorey replied:
When this feature was originally put in the new ui there was some talk about whether to add empty strings.  We decided to allow them, however, now that we have prompt on launch and workflow prompting work in the new ui I believe that adding empty strings in the default values, or in the choices themselves is bad design because it is very confusing when the user goes to answer the survey questions (imagine a drop down with 1 of the choices being an empty string. That would render as just a blank choice).

marshmalien commented on awx/ui_next/src/screens/Template/Survey/SurveyQuestionForm.jsx:
```diff
@@ -182,58 +191,53 @@ function SurveyQuestionForm({
+                    {t`for more information.`}
+                  </>
+                }
+                isRequired
```
It looks like TextAndCheckboxField only accepts 2 props: label and tooltip. Most of the props we are passing in here are not being consumed by TextAndCheckboxField.

marshmalien commented on awx/ui_next/src/screens/Template/Survey/SurveyQuestionForm.jsx:
```diff
@@ -265,5 +269,4 @@ SurveyQuestionForm.defaultProps = {
 };
-
-export default withI18n()(SurveyQuestionForm);
+export default SurveyQuestionForm;
```
👍 Nice

AlexSCorey committed:
moves and renames new field component. adresses console errors, removes unneeded props adds back isVaid prop on formfield

AlexSCorey committed:
allows for resetting the values in edit mode

akus062381 commented:
- [x] Adding a new item, the cursor is not moving down to the next field.  (Alex: this was done intentionally, so that the user is not having to switch lines before completing their first 1.

- [x] The tooltip link is not opening in a new tab (target: _blank)

- [x] Please add ouia id to the multiple choice input field shown when creating a survey.

- [x] The user can save the prompt wizard for a schedule even when a required field (multiple choice) was not provided
OR trying to launch the template without selecting an option the job fails 
(see: [issue](https://github.com/ansible/awx/issues/10129))
![image](https://user-images.githubusercontent.com/63685497/117203571-2c625c80-adbd-11eb-8715-6311b52c5dce.png)

Steps to reproduce:
Create a required multiselect survey
Create a schedule but don’t specify the answer to the question
Hit save button in prompt modal
Hit save button in schedule form 


- [x] When creating survey questions, making a question with multiple choice single answer, and not marking it required, the user should be able to leave the field for that question-answer blank in the prompt modal when launching a job. But there is no option for the user to leave the answer as null. (see [issue](https://github.com/ansible/awx/issues/10128))

- [x] You have 3 survey questions on a job template, none are marked as required. When launching the template, leaving the answers blank. After the template is launched, in the extra variables box there appears the default answers to those questions, followed by an empty string- but only for the text answer type of survey question (which had no answer filled in when prompted). The answers to the non-required multiple choice question was not returned in the extra variables box. ---
- (@akus062381  was going to look into this to determine if it is an api bug, but it is present on devel)

 We found 4 broken tests that we will be working on fixing.

@AlexSCorey

AlexSCorey commented:
> * [ ]  Adding a new item, the cursor is not moving down to the next field (discussed with @akus062381 and shared that I did not do this intentionally. I want the user to be able to create a new line, and then tab over to the button before entering the value for the new line.  
> * [ ]  The tooltip link is not opening in a new tab (target: _blank)- fixed
> * [ ]  Please add ouia id to the multiple choice input field shown when creating a survey.
> * [ ]  The user can save the prompt wizard for a schedule even when a required field (multiple choice) was not provided
>   OR trying to launch the template without selecting an option the job fails
>   ![image](https://user-images.githubusercontent.com/63685497/117203571-2c625c80-adbd-11eb-8715-6311b52c5dce.png)
> 
> Steps to reproduce:
> Create a required multiselect survey
> Create a schedule but don’t specify the answer to the question
> Hit save button in prompt modal
> Hit save button in schedule form
> 
> * [ ]  When creating survey questions, making a question with multiple choice single answer, and not marking it required, the user should be able to leave the field for that question-answer blank in the prompt modal when launching a job. But there is no option for the user to leave the answer as null.
> * [ ]  You have 3 survey questions on a job template, none are marked as required. When launching the template, leaving the answers blank. After the template is launched, in the extra variables box there appears the default answers to those questions, followed by an empty string- but only for the text answer type of survey question (which had no answer filled in when prompted). The answers to the non-required multiple choice question was not returned in the extra variables box.
> 
> We found 4 broken tests that we will be working on fixing.
> 
> @AlexSCorey

AlexSCorey committed:
adds ouiaId and blank target for tooltip

AlexSCorey committed:
updates strings
